gnucash/gnome/assistant-stock-transaction.cpp

```diff
@@ -0,0 +1,1150 @@
+{
+    return;
+}
+
+static void
+add_error (StringVec& errors, const std::string& format_str,
+           const std::string& arg)
+{
+    gchar *buf = g_strdup_printf (format_str.c_str (), arg.c_str());
+    errors.emplace_back (std::string {buf});
+    g_free (buf);
+}
+
+static void
+check_page (GtkListStore *list, gnc_numeric& debit, gnc_numeric& credit,
+            uint splitfield, Account *acct, GtkWidget *memo, GtkWidget *gae,
+            gnc_commodity *comm, bool ignore_account,
+            std::string page, StringVec& errors)
+{
+    if (splitfield == DISABLED)
+        return;
+
+    const gchar* amtstr;
+    gnc_numeric amount;
+    bool debit_side = (splitfield & ENABLED_DEBIT);
+
+    if (gnc_amount_edit_expr_is_valid (GNC_AMOUNT_EDIT (gae), &amount, true, nullptr))
+    {
+        if (splitfield & ALLOW_ZERO)
+            amtstr = "";
+        else
+        {
+            add_error (errors, _("Amount for %s is missing"), page);
+            amtstr = _("(missing)");
+        }
+    }
+    else
+    {
+        if (!(splitfield & ALLOW_NEGATIVE))
+        {
+            if ((splitfield & ALLOW_ZERO) && gnc_numeric_negative_p (amount))
+                add_error (errors, _("Amount for %s must not be negative."), page);
+            else if (!(splitfield & ALLOW_ZERO) && !gnc_numeric_positive_p (amount))
+                add_error (errors, _("Amount for %s must be positive."), page);
+        }
+        if (gnc_numeric_negative_p (amount))
+        {
+            amount = gnc_numeric_neg (amount);
+            debit_side = !debit_side;
+        }
+        if (debit_side)
+            debit = gnc_numeric_add_fixed (debit, amount);
+        else
+            credit = gnc_numeric_add_fixed (credit, amount);
+        amtstr = xaccPrintAmount (amount, gnc_commodity_print_info (comm, true));
+    }
+
+    auto memostr = gtk_entry_get_text (GTK_ENTRY (memo));
+    const gchar *acctstr;
+
+    if (ignore_account)
+        acctstr = "";
+    else if (acct)
+        acctstr = xaccAccountGetName (acct);
+    else
+    {
+        add_error (errors, _("Account for %s is missing"), page);
+        acctstr = _("(missing)");
+    }
+
+    GtkTreeIter iter;
+    gtk_list_store_append (list, &iter);
+    gtk_list_store_set (list, &iter,
+                        SPLIT_COL_ACCOUNT, acctstr,
+                        SPLIT_COL_MEMO, memostr,
+                        SPLIT_COL_DEBIT, debit_side ? amtstr : "",
+                        SPLIT_COL_CREDIT, !debit_side ? amtstr : "",
+                        -1);
+}
+
+static std::string
+join (std::string s, const StringVec& v, char c)
+{
+    auto add_len = [](int a, std::string b) { return a + 1 + b.size (); };
+    auto add_separator_and_string = [&](auto str) { s += c; s += str; };
+    auto length = std::accumulate (v.begin (), v.end (), 0, add_len);
+    s.reserve (length + s.size ());
+    std::for_each (v.begin(), v.end(), add_separator_and_string);
+    return s;
+}
```

Comment:
As we're juggling algorithms, you could try [std::accumulate](https://en.cppreference.com/w/cpp/algorithm/accumulate) with a custom binary operator instead. That conveys the intent of the action even more. There's an example on that page. Freely translated to your use case:

```
auto newline_fold = [](std::string a, std::string b) {
                         return std::move(a) + '\n' + std::move(b);
                     };
 
     s = std::accumulate(std::next(v.begin()), v.end(),
                                    v[0], // start with first element
                                    newline_fold);
```

This ignores the pre-allocation of the result string `s` though. But that may well be a premature optimization. The strings to accumulate typically aren't that long and there may be optimisations already integrated in the accumulate code for strings in the stl itself. Only profiling will tell.

So for initial code I'd aim for simplicity here and just replace the whole `join`function (which appears to be used only once around line 696) by a single call to std::accumulate and embed the lambda function directly into the call, like so:

```
gtk_label_set_text (GTK_LABEL (info->finish_summary),
                    std::accumulate (errors.begin(), errors.end(),
                                     _("The following errors must be fixed:"),
                                     [](std::string a, std::string b) {
                                         return std::move(a) + '\n' + std::move(b);).c_str());
```

If that's too long, you can still declare that lambda just above the call to gtk_label_set_text and use it's variable name instead. If you need this in more places you can define the operator at the top-level as a function and use that function name instead.

Reply:
👍

Reply:
> So for initial code I'd aim for simplicity here and just replace the whole `join`function (which appears to be used only once around line 696) by a single call to std::accumulate and embed the lambda function directly into the call, like so:
> 
> ```
> gtk_label_set_text (GTK_LABEL (info->finish_summary),
>                     std::accumulate (errors.begin(), errors.end(),
>                                      _("The following errors must be fixed:"),
>                                      [](std::string a, std::string b) {
>                                          return std::move(a) + '\n' + std::move(b);).c_str());
> ```
> 
> If that's too long, you can still declare that lambda just above the call to gtk_label_set_text and use it's variable name instead. If you need this in more places you can define the operator at the top-level as a function and use that function name instead.

Ok. However std::accumulate 'inherits' the seed accumulator type so it can't be a char*:
`std::accumulatr (v.begin(), v.end(), std::string { seed }, newline_fold);`